compiler/noirc_evaluator/src/ssa/opt/independent_subgraph_detection.rs

```diff
@@ -0,0 +1,383 @@
+    fn detect_independent_subgraphs_within_function(&self, function: &Function) -> Vec<SsaReport> {
+        let mut context = Context::default();
+
+        // Go through each block in the function and create a list of sets of ValueIds connected by instructions
+        context.block_queue.push(function.entry_block());
+        while let Some(block) = context.block_queue.pop() {
+            if context.visited_blocks.contains(&block) {
+                continue;
+            }
+            context.visited_blocks.insert(block);
+            context.connect_value_ids_in_block(function, block, &self.functions);
+        }
+
+        // Merge ValueIds into sets, where each original small set of ValueIds is merged with another set if they intersect
+        context.merge_sets();
+
+        let function_parameters = function.parameters();
+        let variable_parameters_and_return_values: HashSet<ValueId> = function_parameters
+            .iter()
+            .chain(function.returns())
+            .filter(|&x| match function.dfg[*x] {
+                Value::NumericConstant { .. } => false, // Constant values don't connect elements and can be reused in different subgraphs, so we need to avoid them
+                _ => true,
+            })
+            .copied()
+            .collect();
+
+        let mut connected_sets_indices: HashSet<usize> = HashSet::new();
+
+        // Go through each parameter and each set and check if the set contains the parameter
+        // If it's the case, then that set doesn't present an issue
+        for parameter_or_return_value in variable_parameters_and_return_values.iter() {
+            for (set_index, final_set) in context.value_sets.iter().enumerate() {
+                if final_set.contains(parameter_or_return_value) {
+                    connected_sets_indices.insert(set_index);
+                }
+            }
+        }
+
+        // All the other sets of variables are independent
+        let disconnected_sets_indices: Vec<usize> =
+            HashSet::from_iter(0..(context.value_sets.len()))
+                .difference(&connected_sets_indices)
+                .copied()
+                .collect();
+
+        let all_brillig_generated_values: HashSet<ValueId> =
+            context.brillig_return_to_argument_map.keys().copied().collect();
+
+        // Go through each disconnected set
+        for set_index in disconnected_sets_indices.iter() {
+            let current_set = &context.value_sets[*set_index];
+
+            // Find brillig-generated values in the set
+            let intersection: Vec<ValueId> =
+                all_brillig_generated_values.intersection(current_set).cloned().collect();
+            if intersection.is_empty() {
+                // This is probably a test and the values are optimized in a weird way
+                continue;
+            }
+
+            // Go through all brillig outputs in the set
+            for brillig_output_in_set in intersection.iter() {
```

Comment:
```suggestion
            let intersection = all_brillig_generated_values.intersection(current_set).copied();

            // Go through all brillig outputs in the set
            for brillig_output_in_set in intersection {
```
- Preferring iterators over collecting to a Vec when possible to avoid allocations
- No need for the continue since the only thing after this would be a loop over the empty intersection

Reply:
Done